### PR TITLE
[Merged by Bors] - feat: mul for DirichletCharacter

### PR DESCRIPTION
**Pull request by mo271**
Multiply Dirichlet characters at a common level and obtain their associated primitive character. It also includes a lemma confirming the primitiveness of the resulting character.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Review by MichaelStollBayreuth — APPROVED** [review]
LGTM.

File: Mathlib/NumberTheory/DirichletCharacter/Basic.lean [hunk 1]
```diff
@@ -225,4 +225,20 @@ lemma primitiveCharacter_one (hn : n ≠ 0) :
       (isPrimitive_def _).1 (1 : DirichletCharacter R n).primitiveCharacter_isPrimitive,
       conductor_one hn]
 
+/-- Primitive character associated to multiplication of Dirichlet characters,
+after changing both levels to the same -/
+noncomputable def mul {m : ℕ} (χ₁ : DirichletCharacter R n) (χ₂ : DirichletCharacter R m) :=
```
[the remark this answers is not in the document]
Comment (mo271) [reply Mathlib/NumberTheory/DirichletCharacter/Basic.lean]: done

File: Mathlib/NumberTheory/DirichletCharacter/Basic.lean [hunk 2]
```diff
@@ -224,6 +224,22 @@ lemma primitiveCharacter_one (hn : n ≠ 0) :
       (isPrimitive_def _).1 (1 : DirichletCharacter R n).primitiveCharacter_isPrimitive,
       conductor_one hn]
 
+/-- Primitive character associated to multiplication of Dirichlet characters,
+after changing both levels to the same -/
+noncomputable def mul {m : ℕ} (χ₁ : DirichletCharacter R n) (χ₂ : DirichletCharacter R m) :
+    DirichletCharacter R
+      (changeLevel (Nat.dvd_lcm_left n m) χ₁ * changeLevel (Nat.dvd_lcm_right n m) χ₂).conductor :=
+  primitiveCharacter _
```
[the remark this answers is not in the document]
Comment (laughinggas) [reply Mathlib/NumberTheory/DirichletCharacter/Basic.lean:232]: I am not really sure. Traditionally, multiplication is only defined for primitive characters and naturally these two definitions would then collide. Going ahead a lot of proofs are for the primitive character as well. I suppose one could define `mul` and `primitive_mul`.

[the remark this answers is not in the document]
Comment (laughinggas) [reply Mathlib/NumberTheory/DirichletCharacter/Basic.lean:232]:
```lean
/-- Dirichlet character associated to multiplication of Dirichlet characters,
after changing both levels to the same -/
noncomputable def mul {m : ℕ} (χ₁ : DirichletCharacter R n) (χ₂ : DirichletCharacter R m) :
    DirichletCharacter R
      (Nat.lcm n m) :=
  changeLevel (Nat.dvd_lcm_left n m) χ₁ * changeLevel (Nat.dvd_lcm_right n m) χ₂

/-- Primitive character associated to multiplication of Dirichlet characters,
after changing both levels to the same -/
noncomputable def primitive_mul {m : ℕ} (χ₁ : DirichletCharacter R n) (χ₂ : DirichletCharacter R m) :
    DirichletCharacter R
      (changeLevel (Nat.dvd_lcm_left n m) χ₁ * changeLevel (Nat.dvd_lcm_right n m) χ₂).conductor :=
  primitiveCharacter (mul χ₁ χ₂)
```
I think this should work

**Review by laughinggas — APPROVED** [review]

File: Mathlib/NumberTheory/DirichletCharacter/Basic.lean (hunk 2 above)
[the remark this answers is not in the document]
Comment (eric-wieser) [reply Mathlib/NumberTheory/DirichletCharacter/Basic.lean:232]: The nice thing about defining `mul` first is that you can write the type more succintly:
```
/-- Primitive character associated to multiplication of Dirichlet characters,
after changing both levels to the same -/
noncomputable def primitive_mul {m : ℕ} (χ₁ : DirichletCharacter R n) (χ₂ : DirichletCharacter R m) :
    DirichletCharacter R (mul χ₁ χ₂).conductor :=
  primitiveCharacter (mul χ₁ χ₂)
```

**Review by eric-wieser — COMMENTED** [review]
Please update the PR title to match the commit conventions, and preferably add a short PR description,

File: Mathlib/NumberTheory/DirichletCharacter/Basic.lean (hunk 2 above)
[the remark this answers is not in the document]
Comment (mo271) [reply Mathlib/NumberTheory/DirichletCharacter/Basic.lean]: Nice, @eric-wieser and @laughinggas! That makes a lot of sense to me.
I adapted the code to reflect the suggestions...

[the remark this answers is not in the document]
Comment (laughinggas) [reply Mathlib/NumberTheory/DirichletCharacter/Basic.lean]: I am happy with that.

**Comment by riccardobrasca** [comment]
Thanks!

bors merge

**Pull request closed without merging**